### PR TITLE
enabling multigpu capability

### PR DESCRIPTION
**Pull request by dkazanc**
- enabling multigpu capability through ASTRA and CCPi-regularisation (testing stage) functionality
- Refactored  wrapper for ASTRA, much more concise implementation.

### CODE REVIEW

File: build/src/Python/tomobar/supp/astraOP.py [hunk 1]
```diff
@@ -0,0 +1,358 @@
+#!/usr/bin/env python3
```
[the remark this answers is not in the document]
Comment (dkazanc) [reply build/src/Python/tomobar/supp/astraOP.py:1]: fixed

File: src/Python/tomobar/supp/astraOP.py [hunk 2]
```diff
@@ -69,335 +69,368 @@ def vec_geom_init3D(angles_rad, DetectorSpacingX, DetectorSpacingY, CenterRotOff
...
-        self.AnglesVec = AnglesVec
-        self.ObjSize = ObjSize
-        if CenterRotOffset is None:
+def _set_gpu_device_index(self):
+        #  set device_projector for GPU and enable the GPU indexing
+        try:
+            self.GPUdevice_index = int(self.device_projector) # get a GPU index
+        except ValueError:
+            self.GPUdevice_index = 0 # set to 0 GPU index by default
+        self.device_projector = 'gpu'
```
[the remark this answers is not in the document]
Comment (dkazanc) [reply src/Python/tomobar/supp/astraOP.py:78]: Why is that? What is the alternative?

[the remark this answers is not in the document]
Comment (kalvdans) [reply src/Python/tomobar/supp/astraOP.py:78]: Make a pure functional helper function that doesn't modify any of its argument, that could be called like this:
```
    self.device_projector, self.gpu_index = parse_device_argument(device_projector)
```
The helper function could be defined like
```
def parse_device_argument(device_int_or_string):
    """Convert a cpu/gpu string or integer gpu number into a tuple."""
    if isinstance(device_int_or_string, int):
        return "gpu", device_int_or_string
    return device_int_or_string, 0
```

[the remark this answers is not in the document]
Comment (dkazanc) [reply src/Python/tomobar/supp/astraOP.py:78]: thanks @kalvdans , this looks like more elegant solution indeed. Will try to incorporate.

**Pull request merged**